Fix sections and section editing

annda opened the pull request:
Previous implementation broke sections.

To reproduce:

- Create a page with a few sections (headlines, tables, etc.) Now all sections have edit buttons and can be edited directly.
- Wrap a few sections in <ifauth>
- The section editor is broken, shows wrong content. Edittable plugin cannot parse tables at all.

New document rendering in this PR fixes those problems.

splitbrain committed:
hide content by diverting the document

This changes the way how content is hidden from the user by diverting
the renderer doc to a sacrificial space. This way all rendering outputs
to this space and is simply thrown away after the fact.

This has two advantages:

1) no separate renderer processes need to be instantiated (less resource
hungry)
2) section edit byte positions stay correct

annda committed:
Special handling of headlines

Solution from wrap plugin

annda committed:
PSR-2 cleanup

5p4k commented:
Thanks for the PR!
I was trying this out and I still had some issues with the section editing (see below).
Would it be possible to adapt the PR to fix this too? If that cannot be done easily I will merge it, so that in the meanwhile we get the ToC and edittable fixed.

---

I used [this docker image](https://hub.docker.com/r/mprasil/dokuwiki/) with the following test page
```
==== Sec 1 ====
abc
<ifauth root>
==== Sec 2 ====
def
</ifauth>

==== Sec 3 ====
ghi
```

With this PR, the edit button for section 2 comes right after the section title; editing section 2 however causes both section 2 and 3 to appear in the editor.
![PR version, page](https://user-images.githubusercontent.com/3787759/63385757-e6bfef00-c3a1-11e9-9aee-c146a7a664d0.png)
![PR version, editor](https://user-images.githubusercontent.com/3787759/63385766-ec1d3980-c3a1-11e9-85db-6369caaf63be.png)

annda commented:
You are right, this needs more work. I will take a look at the problem.

splitbrain committed:
added current state of instruction creation

this is a passing test that probably should not pass.

splitbrain committed:
remove section handling

For now we ignore the handling of sections.

splitbrain committed:
store references in the renderer

The plugin gets reused for multiple renderers, but we need independend
references for each one. So we put them into the meta array of the
renderer.

splitbrain committed:
we don't need to handle the headers ourselves

We can simply add baseonly as allowed and let headers do their own
thing.

splitbrain committed:
we no longer even try to handle section editing

Section editing is only available to users who can write the page. User
who can write the page have access to all data in it, so hiding content
from them with ifauthex is useless in the first place.

For proper handling, we might need to overwrite the xhtml renderer and
tune a lot of internal variables and properties.

splitbrain committed:
handle the toc when hiding sections

annda committed:
Dummy doc should be the same type as the original

Compatibility with different renderers, e.g. qc plugin uses an array

annda committed:
Merge pull request #2 from cosmocode/instructiontest

Another go at fixing this plugin

annda commented:
We have changed how the sections are hidden, so there is no data pollution across renderers.

What we have not completely fixed yet are the edit buttons, so will see buttons for hidden sections. However, section editing is only available to users who can write the page. Users
who can write the page have access to all data in it, so preventing them from editing with ifauthex is not a good idea in the first place.

For proper handling, we might need to overwrite the xhtml renderer and tune a lot of internal variables and properties.

Also, a page's table of contents is rendered correctly in the default template, but might include the hidden sections in others, e.g. Sprintdoc.

Take a look and decide if you want to merge this.
